Adding description about mirroring integrations (#211)
demisto/content-docs: docs/integrations/mirroring.md

- `get-mapping-fields` - this command will pull the different incident types and their associated incident fields from the remote system, it will be used in the Mirror out configuration page in the XSOAR UI.

## Useful objects to use in your code
All the functions explained here are globally available through the CommonServerPython file.

### get-remote-data
* GetRemoteDateArgs - this is an object created to maintain all the argument you receive from the server in order to use this command.
Arguments explanation:
  - remote_incident_id - the remote incident id.
  - last_update - the time the incident was last updated - our recommendation would be to use the arg_to_timestamp function to parse it into a timestamp.
* GetRemoteDataResponse - this is the object that maintains the format in which you should order the results from this function. You should use return_results on this object to make it work.
Arguments explanation:
  - mirrored_object - this is essentially the object(dict) of whatever you are trying to mirror - incident in most cases.
  - entries - a list of entries to add to your incident - for instance:
```buildoutcfg
{

>     'Type': entry.get('type'),

Reviewer: this is example is not clear at all.
Provide full example of `get-remote-data` with hard coded values. Don't use constants from XSOAR mirroring, because the reader don't see the full code.
Instead set hard coded values, this way the reader will be able to know what kind/types he should use/pass/return

Arguments explanation:
  - data - will represent the data of the current incident - a dictionary object `{key: value}`
  - entries - will represent the entries from your current incident - a list of dictionaries objects representing the entries
  - remote_incident_id - the remote incident id - string, the id of the incident.
  - inc_status - the status of the incident(numeric value, could be used with IncidentStatus from CommonServerPython)
  - delta - will represent the dictionary of fields which have changed from the last update - a dictionary object `{key: value}` containing only the changed fields.

### get-mapping-fields
* SchemeTypeMapping - this is the object you should use to keep the correct structure of the mapping for the fetched incident type.
* GetMappingFieldsResponse - this is the object to gather all your SchemeTypeMapping and then to parse them properly into the results using the return_results command.
Usage example:
```python
def get_mapping_fields_command():
    xdr_incident_type_scheme = SchemeTypeMapping(type_name=XDR_INCIDENT_TYPE_NAME)

>     for field in XDR_INCIDENT_FIELDS:

Reviewer: same here as my previous note

Arguments explanation:
  - mirrored_object - this is essentially the object(dict) of whatever you are trying to mirror - incident in most cases.
  - entries - a list of entries to add to your incident.
  
An example for such a function could be:
```python
def get_remote_data_command(client, args):
    parsed_args = GetRemoteDateArgs(args)
    new_incident_data: Dict = client.get_incident_data(parsed_args.remote_incident_id, parsed_args.last_update)    new_incident_data: Dict = client.get_incident_data(parsed_args.remote_incident_id, parsed_args.last_update)
    raw_entries: List[dict] = client.get_incident_entries(parsed_args.remote_incident_id, parsed_args.last_update)
    parsed_entries = []
    for entry in raw_entries:
        parsed_entries.append({       
            'Type': entry.get('type'),
            'Category': entry.get('category'),
            'Contents': entry.get('contents'),
            'ContentsFormat': entry.get('format'),
            'Tags': entry.get('tags'),  # the list of tags to add to the entry
            'Note': entry.get('note')  # boolean, True for Note, False otherwise
        })
    
    remote_incident_id = new_incident_data['incident_id']

>     new_incident_data['id'] = remote_incident_id

Author: No, it's only in case the id isn't called id in the remote system

---
id: mirroring_integration
title: Mirroring Integration
---

Server version 6.0.0 adds support for Mirroring Integrations. These integrations allow you to mirror incidents/tickets/cases from other services.  
For example: Another XSOAR server, Cortex XDR, ServiceNow

An example Mirroring Integration can be seen [here](https://github.com/demisto/content/tree/master/Packs/XSOARMirroring/Integrations/XSOARMirroring).

Mirroring Integrations are developed the same as other Integrations. They provide a few extra configuration parameters and APIs.

## Supported Server Version
Mirroring Integration's YAML file _must_ have the following field `fromversion: 6.0.0`. This is because Mirroring Integrations are only supported from Server version 6.0.0 and onwards.

## Required Parameters
Mirroring integration should have the following parameters in the integration YAML file(under the scripts section):
```yml
  isfetch: true
  ismappable: true
  isremotesyncin: true
  isremotesyncout: true
```

> While the `ismappable` and `isremotesyncin` are used for getting information from a remote incident, And the `isremotesyncout` for updating remote data.

Reviewer 2: ```suggestion
Where `ismappable` and `isremotesyncin` are used for getting information from a remote incident, and `isremotesyncout` are used for updating remote data.
```

            'Tags': entry.get('tags'),  # the list of tags to add to the entry
            'Note': entry.get('note')  # boolean, True for Note, False otherwise
        })
    
    remote_incident_id = new_incident_data['incident_id']
    new_incident_data['id'] = remote_incident_id
    return GetRemoteDataResponse(new_incident_data, parsed_entries)
```

### update-remote-system
* UpdateRemoteSystemArgs - this is an object created to maintain all the arguments you receive from the server in order to use this command.
Arguments explanation:
  - data - will represent the data of the current incident - a dictionary object `{key: value}`
  - entries - will represent the entries from your current incident - a list of dictionaries objects representing the entries
  - remote_incident_id - the remote incident id - string, the id of the incident.
  - inc_status - the status of the incident(numeric value, could be used with IncidentStatus from CommonServerPython)
  - delta - will represent the dictionary of fields which have changed from the last update - a dictionary object `{key: value}` containing only the changed fields.
  
An example for such a function could be:

> ```

Reviewer 2: ```suggestion
```python
```

        demisto.debug(f'Closing remote incident {new_incident_id}')
        client.close_incident(
            new_incident_id,
            updated_incident.get('version'),  # type: ignore
            parsed_args.data.get('closeReason'),
            parsed_args.data.get('closeNotes')
        )

    return new_incident_id

```
### get-mapping-fields
* SchemeTypeMapping - this is the object you should use to keep the correct structure of the mapping for the fetched incident type.
* GetMappingFieldsResponse - this is the object to gather all your SchemeTypeMapping and then to parse them properly into the results using the return_results command.
Usage example:
```python
def get_mapping_fields_command():
    xdr_incident_type_scheme = SchemeTypeMapping(type_name='incident type example')
    for field in ['field1', 'field2', 'field3']:
        xdr_incident_type_scheme.add_field(name=field, description='the description for the field')

    return GetMappingFieldsResponse(xdr_incident_type_scheme)

```

## Incident fields on a XSOAR incident 
There are few incident fields you need to configure to get going on the system.

> * dbotMirrorDirection - Both or In

Author: ```suggestion
* dbotMirrorDirection - Both, In, out or None
```

* dbotMirrorId - represents the id of the incident in the external system
* dbotMirrorInstance - the instance that will perform the mirroring

Useful fields:
* dbotMirrorTags - tags for mirrored-out entries (comment/files).

## Debugging

* getMirrorStatistics command - a hidden command that returns mirroring statistics: total mirrored, rate limited, and last run.
* getSyncMirrorRecords - a hidden command that returns records that hold internal mirroring metadata for each mirrored incident
* get-remote-data - is runnable through the war room with the relevant arguments and see the results.
* get-mapping-fields - is runnable through the war room with no arguments and see the results.

Also important to note:

> DO NOT change dbot mirroring fields for existing incidents after they were set

Reviewer 2: need more clarity here
why would someone want to change these fields? what else should they do ?

---
id: mirroring_integration
title: Mirroring Integration
---

> Server version 6.0.0 adds support for Mirroring Integrations. These integrations allow you to mirror incidents/tickets/cases from other services for example: Another XSOAR server, Cortex XDR, ServiceNow

Reviewer: ```suggestion
Server version 6.0.0 adds support for Mirroring Integrations. These integrations allow you to mirror incidents/tickets/cases from other services.   
For example: another XSOAR server, Cortex XDR, ServiceNow
```

An example Mirroring Integration can be seen [here](https://github.com/demisto/content/tree/master/Packs/XSOARMirroring/Integrations/XSOARMirroring).

Mirroring Integrations are developed the same as other Integrations. They provide a few extra configuration parameters and APIs.

## Supported Server Version
Mirroring Integration's YAML file _must_ have the following field `fromversion: 6.0.0`. This is because Mirroring Integrations are only supported from Server version 6.0.0 and onwards.

## Required Parameters

> Mirroring integration should have the following parameters in the integration YAML file:

Reviewer 3: These params need to be under `script` in the yml right? like `isfetch` right now?

```yml
  isfetch: true
  ismappable: true
  isremotesyncin: true
  isremotesyncout: true
```
While the `ismappable` and `isremotesyncin` are used for getting information from a remote incident, And the `isremotesyncout` for updating remote data.

## Commands

> Mirroring Integration should have some and at some cases all(When using full mirror in and out) of the following commands:

Reviewer 3: ```suggestion
Use the following commands to implement a mirroring integration. 
*Note that when mirroring both incoming and outgoing data all the commands are required, while only some of the commands are required for a one-way mirror.*
```

- `test-module` - this is the command that is run when the `Test` button in the configuration panel of an integration is clicked.
- `fetch-incidents` - this is the command that fetches new incidents to Cortex XSOAR.
- `get-remote-data` - this command takes new information about the incidents in the remote system and updates the *already existing* incidents in Cortex XSOAR. This command is being executed every 1 minute. 
- `update-remote-system` - this command updates the remote system with the information we have in the mirrored incidents within Cortex XSOAR. This command is being executed every 1 minute.

> - `get-mapping-fields` - this command will pull the different incident types and their associated incident fields from the remote system, it will be used in the Mirror out configuration page in the XSOAR UI.

Reviewer 3: is there a doc about the mirror out configuration page? - if so we should link it

Author: not sure, @yaron-libman maybe you know

## Usefull objects to use in your code

> All the functions explained here are globally available through the CommonServerPython file.

Reviewer 3: Maybe add a link to the code for every object? might be easier to understand and know all the internal functions for them

Author: this will change too often and will not be maintainable

### get-remote-data

> * GetRemoteDateArgs - this is an object created to maintain all the argument you receive from the server in order to use this command.

Reviewer 3: We should elaborate on what args are available.
Currently, it is only `id` and `last_update` but it might expand in the future and here is a good point to make sure everyone is up to speed.

Also, we should add more info about how `last_update` is stored - timestamp (milliseconds or seconds, str or int)? datetime object? maybe even point to the `arg_to_timestamp` function for consistency?

> * GetRemoteDataResponse - this this is the object that maintains the format in which you should order the results from this function. You should use return_results on this object to make it work.

Reviewer: ```suggestion
* GetRemoteDataResponse - this is the object that maintains the format in which you should order the results from this function. You should use return_results on this object to make it work.
```

> * GetRemoteDataResponse - this this is the object that maintains the format in which you should order the results from this function. You should use return_results on this object to make it work.

Reviewer 3: We should explain args here as well like what are `entries`? how should you pass them?

### get-mapping-fields

> * UpdateRemoteSystemArgs - this is an object created to maintain all the argument you receive from the server in order to use this command.

Reviewer: I would say describe here all the `UpdateRemoteSystemArgs` args.

---
id: mirroring_integration
title: Mirroring Integration
---

Server version 6.0.0 adds support for Mirroring Integrations. These integrations allow you to mirror incidents/tickets/cases from other services.  
For example: Another XSOAR server, Cortex XDR, ServiceNow

> An example Mirroring Integration can be seen [here](https://github.com/demisto/content/tree/master/Packs/XSOARMirroring/Integrations/XSOARMirroring).

Reviewer 4: This link is broken

Author: it's not in master yet.

- `test-module` - this is the command that is run when the `Test` button in the configuration panel of an integration is clicked.
- `fetch-incidents` - this is the command that fetches new incidents to Cortex XSOAR.
- `get-remote-data` - this command takes new information about the incidents in the remote system and updates the *already existing* incidents in Cortex XSOAR. This command is being executed every 1 minute. 
- `update-remote-system` - this command updates the remote system with the information we have in the mirrored incidents within Cortex XSOAR. This command is being executed every 1 minute.
- `get-mapping-fields` - this command will pull the different incident types and their associated incident fields from the remote system, it will be used in the Mirror out configuration page in the XSOAR UI.

## Useful objects to use in your code
All the functions explained here are globally available through the CommonServerPython file.

### get-remote-data
* GetRemoteDateArgs - this is an object created to maintain all the argument you receive from the server in order to use this command.
Arguments explanation:

>   - remote_incident_it - the remote incident id.

Reviewer 3: ```suggestion
  - remote_incident_id - the remote incident id.
```

>   - remote_incident_it - the remote incident id.

Reviewer 3: ```suggestion
  - remote_incident_id - the remote incident id.
```

  - last_update - the time the incident was last updated - our recommendation would be to use the arg_to_timestamp function to parse it into a timestamp.
* GetRemoteDataResponse - this is the object that maintains the format in which you should order the results from this function. You should use return_results on this object to make it work.
Arguments explanation:
  - mirrored_object - this is essentially the object(dict) of whatever you are trying to mirror - incident in most cases.
  - entries - a list of entries to add to your incident - for instance:
```buildoutcfg
{
    'Type': entry.get('type'),
    'Category': entry.get('category'),
    'Contents': entry.get('contents'),
    'ContentsFormat': entry.get('format'),
    'Tags': entry.get('tags'),  # the list of tags to add to the entry
    'Note': entry.get('note')  # boolean, True for Note, False otherwise
}
```

### update-remote-system
* UpdateRemoteSystemArgs - this is an object created to maintain all the arguments you receive from the server in order to use this command.
Arguments explanation:

>   - data - will represent the data of the current incident.

Reviewer: add examples for each argument here.
So the developer will have better understand what is the structure of entries, delta, what is the possible options he will get from inc_status (number is not enough). What each number represents

  - entries - will represent the entries from your current incident 

>   - remote_incident_it - the remote incident id.

Reviewer 3: ```suggestion
  - remote_incident_id - the remote incident id.
```

  - inc_status - the status of the incident(numeric value, could be used with IncidentStatus from CommonServerPython)
  - delta - will represent the dictionary of fields which have changed from the last update.

> ### update-remote-system

Reviewer: ```suggestion
### get-mapping-fields
```

* SchemeTypeMapping - this is the object you should use to keep the correct structure of the mapping for the fetched incident type.

> * GetMappingFieldsResponse - this is the object to gather all your SchemeTypeMapping and then to parse them properly into the results using the return_results command.

Reviewer: add an example of how to construct `GetMappingFieldsResponse` via `SchemeTypeMapping` and return it

## ￿Get going on XSOA￿R incident
There are few incident fields you need to configure to get going on the system.
* dbotMirrorDirection - Both or In
* dbotMirrorId - represents the id of the incident in the external system
* dbotMirrorInstance - the instance that will perform the mirroring

Useful fields:
* dbotMirrorTags - tags for mirrored-out entries (comment/files).

## Debugging

* getMirrorStatistics command - a hidden command that returns mirroring statistics: total mirrored, rate limited, and last run.

> * getSyncMirrorRecords - a hidden command that returns records that hold internal mirroring metadata for each mirrored incident

Reviewer: explain how to debug/troubleshoot get-remote-data and update-remote-system commands.
